### PR TITLE
Implement P2419R2 Clarify Handling Of Encodings In Localized Formatting Of `chrono` Types

### LINKED ISSUE #2927: P2419R2 Clarify Handling Of Encodings In Localized Formatting Of `chrono` Types
[P2419R2](https://wg21.link/P2419R2) Clarify Handling Of Encodings In Localized Formatting Of `chrono` Types

**Increased** feature-test macro (expected):
```cpp
#define __cpp_lib_format 202207L
```

The feature-test macro bump is shared with #2937, so don't update until both are implemented.

### PR DESCRIPTION
**Pull request by cpplearner**
Fixes #2927.

When the literal encoding is UTF-8, This ensures that the strings produced by chrono formatters are always encoded in UTF-8, by first formatting with `wchar_t`, then converting the output to UTF-8. (Is there a better way?)

I didn't convert the output of `_Custom_write`, because it seems that `_Custom_write` generally does not deal with localized formatting.

### CODE REVIEW

**Comment by barcharcraz** [comment]
does this need a __cpp_lib_format bump?

File: tests/std/tests/P0355R7_calendars_and_time_zones_formatting/test.cpp
Comment (barcharcraz) [inline tests/std/tests/P0355R7_calendars_and_time_zones_formatting/test.cpp:1122]: should we be running these if the literal encoding is _not_ utf-8?

File: stl/inc/chrono
Comment (barcharcraz) [reply stl/inc/chrono:13]: oh, it's because this is a move from filesystem. Ugh

File: stl/inc/chrono
Comment (barcharcraz) [reply stl/inc/chrono:100]: oh, actually it might just pull the utf-16 version of the locale. That's probably fine.

File: tests/std/tests/P0355R7_calendars_and_time_zones_formatting/test.cpp
Comment (cpplearner) [reply tests/std/tests/P0355R7_calendars_and_time_zones_formatting/test.cpp:1122]: The test is only compiled with `/utf-8`.

https://github.com/microsoft/STL/blob/ba73f5a60a7a650df0f9d54f9710c33bb1c0d0b2/tests/std/tests/P0355R7_calendars_and_time_zones_formatting/env.lst#L4-L6

We can test that, if the literal encoding is not utf-8, `format(locale{"zh-CN"}, STR("{:^22L%Y %B %d %A}"), 2021y / June / 16d)` produces `" 2021 \xC1\xF9\xD4\xC2 16 \xD0\xC7\xC6\xDA\xB6\xFE  "` (text is GB2312-encoded) after `setlocale(LC_ALL, ".936")` and produces `" 2021 \xE5\x85\xAD\xE6\x9C\x88 16 \xE6\x98\x9F\xE6\x9C\x9F\xE4\xBA\x8C  "` (text is UTF8-encoded) after `setlocale(LC_ALL, ".UTF-8")`, but I feel that it isn't worth the trouble.

**Comment by cpplearner** [comment]
> does this need a __cpp_lib_format bump?

[P2508R1](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2508r1.html) (Expose std::*basic-format-string*) updates the same macro. I'll wait until that paper is implemented.

File: stl/inc/filesystem
Comment (strega-nil-ms) [inline stl/inc/filesystem:23]: I don't like removing these `include`s; no, they're not necessary anymore because they're gotten through chrono, but we still use them in this file.

File: stl/inc/chrono
Comment (strega-nil-ms) [inline stl/inc/chrono:5432]: I don't know why testing didn't catch it, but `view` is only available in C++20 and later, and this is a C++17 feature.

File: stl/inc/chrono
Comment (strega-nil-ms) [inline stl/inc/chrono:5424]: Can we use `_wcsftime_l` instead? I might just hate iostreams, but I would personally prefer to use the C standard library directly here...

File: tests/std/tests/P0355R7_calendars_and_time_zones_formatting/test.cpp
Comment (strega-nil-ms) [reply tests/std/tests/P0355R7_calendars_and_time_zones_formatting/test.cpp:1122]: It would be nice if this was checked for C++17 as well.

File: stl/inc/chrono
Comment (barcharcraz) [reply stl/inc/chrono:5432]: chrono formatting is C++20 only. The C++17 stuff in this header is just moving some c++17 stuff from filesystem into chrono.

File: stl/inc/chrono
Comment (barcharcraz) [reply stl/inc/chrono:5424]: it's to match what the formatter was doing below, and to match how chrono formatting is specified.

File: stl/inc/filesystem
Comment (barcharcraz) [reply stl/inc/filesystem:23]: I agree, let's keep them.

Comment (cpplearner) [reply stl/inc/filesystem:23]: I'd prefer not to pay the cost of including `<string_view>`. It's not transitively included in C++17 mode and is not needed, since it simply checks `_HAS_CXX17` (which `<filesystem>` already checks) and includes `<xstring>` (which `<filesystem>` already includes).

It might make sense to include `<system_error>`, `<xfilesystem_abi.h>` and `<xstring>` and avoid transitive inclusion, but `<filesystem>` relies on transitive inclusion anyway. For example, It uses `iterator_traits` without directly including `<iterator>` or `<__msvc_iter_core.h>`.

Comment (strega-nil-ms) [reply stl/inc/filesystem:23]: You can include `<xstring>` instead if you want, but the cost of including a header twice is minimal - this is something that compilers have optimized _a lot_.

**Comment by StephanTLavavej** [comment]
@cpplearner Just checking on the status of this PR - are you planning to make changes here, or should we mark it as blocked until there's a PR for #2937?

**Comment by cpplearner** [comment]
I thought I made the changes but I forgot to push them 😢

File: tests/std/tests/P0355R7_calendars_and_time_zones_formatting/test.cpp
Comment (cpplearner) [reply tests/std/tests/P0355R7_calendars_and_time_zones_formatting/test.cpp:1122]: Chrono formatting is C++20-only.

**Comment by StephanTLavavej** [comment]
@cpplearner Checking again - are your changes ready to push? @barcharcraz and @strega-nil-ms would like to review them :smile_cat:

**Comment by cpplearner** [comment]
I've pushed db9300996bc8cfca932eab9e44d5165038abf756 after posting https://github.com/microsoft/STL/pull/2977#issuecomment-1211400036 (GitHub UI is weird). I think this PR is ready for review now.

File: stl/inc/filesystem
Comment (strega-nil-ms) [reply stl/inc/filesystem:23]: oh oops I just read what you said, nvm.

**Review by barcharcraz — APPROVED** [review]
This now looks good.

I'm still a tiny bit annoyed with the approach of narrowing the wide versions of the locale data, but I can't think of a better way to do it. Maybe I'll make an issue to switch to the UTF-8 locale data once we no longer support systems that lack it.

**Review by strega-nil-ms — APPROVED** [review]
LGTM! Thanks, although I'm kinda sad for using wostringstream.

File: stl/inc/yvals_core.h
Comment (StephanTLavavej) [inline stl/inc/yvals_core.h:70]: This is specific to the behavior of C++20 chrono formatting. Therefore it should be listed under `_HAS_CXX20`, not the truly unconditional (i.e. C++14) section.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
LGTM too, thanks @cpplearner! FYI @barcharcraz @strega-nil-ms I pushed a one-line change after you approved, moving the `yvals_core.h` comment to the C++20 section as this is specific to chronat.

**Comment by StephanTLavavej** [comment]
Thanks for implementing this feature - just in *time* for 17.4 Preview 3! :joy_cat: :stopwatch: :tada:

**Pull request merged**
